### PR TITLE
Script for generating changelogs, new template for PR's

### PR DESCRIPTION
**Pull request by igordmn**

### CODE REVIEW

**Comment by igordmn** [comment]
Keeping it as a draft until it proved itself useful

File: tools/changelog.main.kts [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+                    .groupBy { it.subsectionName() }
+                    .forEach { (subsection, subsectionEntries) ->
+                        appendLine("### $subsection")
+                        subsectionEntries.forEach {
+                            appendLine(it.format())
+                        }
+                        appendLine()
+                    }
+            }
+    }
```
Comment (MatkovIvan) [inline tools/changelog.main.kts:88]: I guess we can parse configs and automatically add

```suggestion
        appendLine("## Dependencies\n")
        appendLine("- Gradle Plugin `org.jetbrains.compose`, version `${version}`. Based on Jetpack Compose libraries:")
        // TBD
        appendLine("___")
    }
```

File: tools/changelog.main.kts [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+println(
+    buildString {
+        append("_Changes since ${commitToVersion(firstCommit)}_\n")
+        append("\n")
+
+        entries
+            .sortedBy { it.sectionOrder() }
+            .groupBy { it.sectionName() }
+            .forEach { (section, sectionEntries) ->
+                appendLine("## $section")
```
[the remark this answers is not in the document]
Comment (igordmn) [reply tools/changelog.main.kts:108]: Fixed

File: tools/changelog.main.kts [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+ * Transforms v1.6.0-beta01 to 1.6.0-beta01
+ */
+fun commitToVersion(commit: String) =
+    if (commit.startsWith("v") && commit.contains(".")) {
+        commit.removePrefix("v")
+    } else {
+        commit
+    }
+
+fun ChangelogEntry.format() = if (link != null) "$message ([link]($link))" else message
```
[the remark this answers is not in the document]
Comment (igordmn) [reply tools/changelog.main.kts]: Fixed

File: tools/changelog.main.kts [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+)
+
+if (token == null) {
+    println("To increase the rate limit, specify token (https://github.com/settings/tokens): kotlin changelog.main.kts <firstCommit> <lastCommit> TOKEN")
+}
+
+val entries = entriesForRepo("JetBrains/compose-multiplatform-core") +
+        entriesForRepo("JetBrains/compose-multiplatform")
+
+println("\n# CHANGELOG")
```
[the remark this answers is not in the document]
Comment (igordmn) [reply tools/changelog.main.kts]: Fixed

File: tools/changelog.main.kts (hunk 1 above)
Comment (igordmn) [reply tools/changelog.main.kts:145]: Fixed

**Comment by igordmn** [comment]
New features in the last commits:
- exclude released cherry-picks
- include non-merged cherry-picked PR's
- change CHANGELOG.md instead of printing
- read the previous version from CHANGELOG.md
- set the same formatting as in the final changelog

**Comment by igordmn** [comment]
@MatkovIvan, it is ready for merge, could you review it now?

File: tools/changelog.main.kts [hunk 5]
```diff
@@ -0,0 +1,416 @@
...
+ * Converts:
+ * Message (title)[some link], message
+ *
+ * to:
+ * Message title, message
+ */
+fun String.removeLinks(): String = replace(Regex("\\[([^)]*)\\]\\([^\\]]*\\)"), "$1")
+
+/**
+ * Extract by format https://github.com/JetBrains/compose-multiplatform/blob/b32350459acceb9cca6b9e4422b7aaa051d9ae7d/.github/PULL_REQUEST_TEMPLATE.md?plain=1
```
Comment (MatkovIvan) [inline tools/changelog.main.kts:211]: Should we refer `jb-main` instead `b32350459acceb9cca6b9e4422b7aaa051d9ae7d`?

File: tools/changelog.main.kts [hunk 6]
```diff
@@ -0,0 +1,416 @@
...
+    val list = mutableListOf<ChangelogEntry>()
+    var section: String? = null
+    var subsection: String? = null
+
+    for (line in relNoteBody.orEmpty().split("\n")) {
+        // parse "### Section - Subsection"
+        if (line.startsWith("### ")) {
+            val s = line.removePrefix("### ")
+            section = s.substringBefore("-", "").trim().ifEmpty { null }
+            subsection = s.substringAfter("-", "").trim().ifEmpty { null }
```
Comment (MatkovIvan) [inline tools/changelog.main.kts:237]: Let's print some warning if it's not in the predefined list

**Review by MatkovIvan — APPROVED** [review]

File: tools/changelog.main.kts (hunk 5 above)
Comment (igordmn) [reply tools/changelog.main.kts]: Done

**Pull request merged**